src/dac.rs

```diff
@@ -0,0 +1,220 @@
+//! Configure the internal DAC on the stm32f3xx.
+//! Incomplete, but includes basic operation.
+//!
+//! You should have the approprite dac pin set up as an analog input, to prevent
+//! parasitic power consumption. For example:
+//! ```rust
+//! let _dac1_pin = gpioa.pa4.into_analog(&mut gpioa.moder, &mut gpioa.pupdr);
+//! let _dac2_pin = gpioa.pa5.into_analog(&mut gpioa.moder, &mut gpioa.pupdr);
+//! ```
+//!
+
+// todo: DAC trait that will go in a PR to embeddd hal.
+// [PR](https://github.com/rust-embedded/embedded-hal/pull/247)
+
+//! Digital-analog conversion traits
+//!
+//! A trait used to identify a digital-to-analog converter, and its
+//! most fundamental features.
+
+use core::fmt;
+
+use crate::{pac, rcc::APB1};
+
+pub trait SingleChannelDac {
+    /// Error type returned by DAC methods
+    type Error;
+    type Word;
+
+    /// Output a constant signal, given a bit word.
+    fn try_set_value(&mut self, value: Self::Word) -> Result<(), Self::Error>;
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum DacId {
+    One,
+    Two,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum DacBits {
+    EightR,
+    TwelveL,
+    TwelveR,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum Trigger {
+    Tim6,
+    Tim3_8,
+    Tim7,
+    Tim15,
+    Tim2,
+    Tim4, // Not available on DAC 2.
+    Exti9,
+    Swtrig,
+}
```

Comment:
Instead of
```Rust
pub enum Trigger {
  /// Timer 15
  Tim15
  /// A software trigger
  Swtrig
}
```
I'd love to see simply
```
pub Trigger {
  Timer15,
  SoftwareTrigger,
}
```
or something alike.
I also see the need to at least document that Timer 4 is not available on DAC 2. Even better would be to not allow it during compilation.

Reply:
Since this PR doesn't include support for DAC2 anymore, I don't think it is necessary to specifically mention that Timer 4 is not available there. That would just confuse people.

I originally requested the per-variant annotations to document that Timer 4 edge-case. Since that now isn't relevant anymore, I agree that the comments don't add much value anymore either.